### PR TITLE
network: Setup DNS for sandbox

### PR DESCRIPTION
**Pull request by nitkon**
Sandbox has network but not dns. Set
dns in the sandbox.

Fixes: #589

Signed-off-by: Nitesh Konkar <niteshkonkar@in.ibm.com>

### CODE REVIEW

**Comment by GabyCT** [comment]
@nitkon , sorry but I am kind of confussed there are 3 PRs in this repository which has the same title and modify the same files, which one is the one that we need to check? thanks

**Comment by nitkon** [comment]
@GabyCT, @amshinde, @bergwolf: Please review this PR. Thnx.

**Comment by devimc** [comment]
@nitkon could you close the other PRs?

File: network.go [hunk 1]
```diff
@@ -575,10 +577,27 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
 // DNS //
 /////////
 
-func setupDNS(dns []string) error {
+func writeDNSFile(lines []string, path string) (err error) {
```
Comment (devimc) [inline network.go]: since this function is only used in `setupDNS`, please rename it to `setupDNS`

**Comment by nitkon** [comment]
@devimc: Updated this PR and closed other PR's. Ptal ...

File: network.go [hunk 2]
```diff
@@ -575,7 +576,21 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
 // DNS //
 /////////
 
-func setupDNS(dns []string) error {
+func setupDNS(dns []string) (err error) {
+	var file *os.File
+	GuestDNSFile := "/etc/resolv.conf"
```
Comment (devimc) [inline network.go]: this variable should be declared outside this function and not exported (the first character shouldn't be a capital letter" `guestDNSFile`), that way you can write a unit test

File: network_test.go [hunk 3]
```diff
@@ -322,3 +324,18 @@ func TestListRoutes(t *testing.T) {
...
+	guestDNSFile = tmpfile.Name()
+	defer os.RemoveAll(guestDNSFile)
+
+	dns := []string{
+		"nameserver 8.8.8.8",
+		"nameserver 8.8.4.4",
+	}
+
+	err = setupDNS(dns)
+	assert.Nil(t, err)
```
Comment (amshinde) [inline network_test.go]: Test looks good. Would be better if you check the contents of the file are actually what you expect to see.

File: network.go [hunk 4]
```diff
@@ -575,7 +577,20 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
 // DNS //
 /////////
 
-func setupDNS(dns []string) error {
+func setupDNS(dns []string) (err error) {
```
Comment (amshinde) [inline network.go]: nit: So looks like you are now sending the dns string from the runtime itself. Can you mention that in the commit as well.

File: network_test.go [hunk 5]
```diff
@@ -322,3 +324,18 @@ func TestListRoutes(t *testing.T) {
 	assert.True(reflect.DeepEqual(results.Routes[1], inputRoutesSimple[1]),
 		"Route listed didn't match: got %+v, expecting %+v", results.Routes[1], inputRoutesSimple[1])
 }
+
+func TestSetupDNS(t *testing.T) {
+	tmpfile, err := ioutil.TempFile("", "resolv.conf")
+	assert.Nil(t, err)
```
Comment (devimc) [inline network_test.go]: `assert.NoError(t, err)`

File: network_test.go (hunk 3 above)
Comment (devimc) [inline network_test.go]: `assert.NoError(t, err)`

**Comment by nitkon** [comment]
Updated PTAL @devimc @amshinde 
/test

**Comment by devimc** [comment]
@nitkon 

```
Stderr: rpc error: code = Unknown desc = open /etc/resolv.conf: read-only file system
```

oops! we are running on a read-only rootfs, but you can bind-remount this file as a read-write

**Comment by nitkon** [comment]
@devimc : Since the rootfs is readonly, the file creation of `/etc/resolv.conf` would have failed? If file creation failed, which file should be bind-remount as RW ?

**Comment by devimc** [comment]
@nitkon ahh ok, then you should create it and then bind-remount it, https://github.com/kata-containers/osbuilder/blob/master/image-builder/image_builder.sh#L391

**Comment by jcvenegas** [comment]
Hey @nitkon , just comming from https://github.com/kata-containers/osbuilder/pull/346 what is the main use case for this PR? I dont fully get the relation with the issue https://github.com/kata-containers/agent/issues/589 (probably wrong reference? )

File: network.go (hunk 4 above)
Comment (jcvenegas) [reply network.go]: I dont see where this function is being called, could you point me to it ?

**Comment by nitkon** [comment]
@jcvenegas : Looks like a wrong reference. The main aim of this PR is to solve this [issue](https://github.com/kata-containers/runtime/issues/1603). I shall create a similar issue in the agent repo and will reference it in my next update to this PR.

File: network.go (hunk 4 above)
Comment (jcvenegas) [reply network.go]: Re: sorry I did not see is a replacement the function, it is not a new called function.

**Comment by jcvenegas** [comment]
@nitkon that reference has more sense :) , if we decide to go with this PR yeah lets create new issue, meanwhile lets talk about it in the original issue.

**Comment by nitkon** [comment]
@amshinde @devimc @bergwolf  : Updated my PR, Ptal ...

**Comment by nitkon** [comment]
Hi @devimc, I am testing if the file contents are as expected. This is because Travis fails to mount file with permission denied. 
```
=== RUN   TestSetupDNS

--- FAIL: TestSetupDNS (0.00s)

    Error Trace:    network_test.go:447

    Error:          Received unexpected error:

                    rpc error: code = Internal desc = Could not mount /tmp/resolv.conf142278078 to /tmp/resolv.conf425331995: operation not permitted
```

I logged into Travis to figure out whats happening and realized probably we need to run the test as sudo.

```
travis@travis-20190814101414:~/gopath/src/github.com/kata-containers/agent$ go test -v -run TestSetupDNS                                                      
=== RUN   TestSetupDNS
--- FAIL: TestSetupDNS (0.00s)
        Error Trace:    network_test.go:447
        Error:          Received unexpected error:
                        rpc error: code = Internal desc = Could not mount /tmp/resolv.conf724125373 to /tmp/resolv.conf674219862: operation not permitted
        Test:           TestSetupDNS
        Error Trace:    network_test.go:453
        Error:          Not equal: 
                        expected: []string{"nameserver 8.8.8.8", "nameserver 8.8.4.4"}
                        actual  : []string{""}
                    
                        Diff:
                        --- Expected
                        +++ Actual
                        @@ -1,4 +1,3 @@
                        -([]string) (len=2) {
                        - (string) (len=18) "nameserver 8.8.8.8",
                        - (string) (len=18) "nameserver 8.8.4.4"
                        +([]string) (len=1) {
                        + (string) ""
                         }
        Test:           TestSetupDNS
FAIL
exit status 1
FAIL    github.com/kata-containers/agent        0.009s


travis@travis-20190814101414:~/gopath/src/github.com/kata-containers/agent$ which go
/home/travis/.gimme/versions/go1.12.9.linux.ppc64le/bin/go
travis@travis-20190814101414:~/gopath/src/github.com/kata-containers/agent$ sudo cat /etc/sudoers
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/home/travis/.gimme/versions/go1.12.9.linux.ppc64le/bin/"
root    ALL=(ALL:ALL) ALL
#includedir /etc/sudoers.d



travis@travis-20190814101414:~/gopath/src/github.com/kata-containers/agent$ sudo go test -v -run TestSetupDNS                                                 
=== RUN   TestSetupDNS
--- PASS: TestSetupDNS (0.00s)
PASS
ok      github.com/kata-containers/agent        0.014s
travis@travis-20190814101414:~/gopath/src/github.com/kata-containers/agent$
```

Should we run this particular test as sudo or we have a cleaner method. 
cc @grahamwhaley

**Comment by nitkon** [comment]
@devimc : Ptal ... 
/retest

**Comment by jodh-intel** [comment]
Any more comments @amshinde?

File: network.go [hunk 6]
```diff
@@ -609,8 +612,27 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
...
+			_, err = file.WriteString(strings.TrimSpace(line) + "\n")
+		} else {
+			_, err = file.WriteString(strings.TrimSpace(line))
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	err = mount(kataGuestSandboxDNSFile, guestDNSFile, "bind", syscall.MS_BIND, "")
```
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: @nitkon Can you address this? You should just use `return mount(kataGuestSandboxDNSFile, guestDNSFile, "bind", syscall.MS_BIND, "")` here.

File: network.go [hunk 7]
```diff
@@ -609,8 +612,31 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
...
+		agentLog.Debug("Did not set sandbox DNS as DNS not received as part of grpc request.")
+		return nil
+	}
+	if file, err = os.Create(kataGuestSandboxDNSFile); err != nil {
+		return err
+	}
+	defer file.Close()
+
+	for i, line := range dns {
+		if i != (len(dns) - 1) {
```
Comment (amshinde) [inline network.go]: nit: Its better to address the positive case first with reversing the order of the conditions.

**Comment by nitkon** [comment]
@amshinde: Addressed the review comments.
/test

**Comment by nitkon** [comment]
All checks have passed.

**Pull request merged**